server/src/Icepeak/Server/WebsocketServer.hs

```diff
@@ -36,6 +45,43 @@ import qualified Icepeak.Server.Metrics as Metrics
 import qualified Icepeak.Server.Subscription as Subscription
 import Data.Maybe (isJust)
 
+-- | 'WS.ServerApp' parameterized over the last received pong timestamp. See
+-- 'WSServerOptions'.
+type WSServerApp = WSServerOptions -> WS.ServerApp
+
+-- | Any additional payload information information used by the websocket
+-- application.
+--
+-- Currently this is only used to keep track of the last received pong. This
+-- value is initialized to the current time when starting the application, and
+-- it is updated in a connection-specific @connectionOnPong@ handler. When a
+-- ping is sent, the IO action in 'withInterruptiblePingThread' checks whether
+-- the client answered our last ping with a pong. If it hasn't, then the server
+-- will terminate the websocket connection as timeouts would otherwise leave
+-- zombie connections.
+newtype WSServerOptions = WSServerOptions { wsOptionLastPongTime :: MVar TimeSpec }
```

Comment:
I don't see any use of the synchronization part of the MVar, wouldn't a regular `IORef` with atomic updates suffice here?

Reply:
I wasn't entirely sure whether the pong would be received on the same thread the ping was sent from, which is why I went with `MVar` to avoid reading partial writes. `TimeSpec` consists of two `Int64`s, so they can never be read or written to atomically without external synchronization.

Reply:
> `TimeSpec` consists of two `Int64`s, so they can never be read or written to atomically without external synchronization.

You're thinking Rust but writing Haskell :slightly_smiling_face:   `IORef`s are always boxed, and thus always represented as a pointer which can be both read and written atomically without external synchronization.
For portable code, that should probably still happen through `atomicModifyIORef`.
Also see https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-IORef.html#g:2

Reply:
Ah yeah good point, shouldn't make any any real difference performance wise but I'll switch this out for an `IORef` since partial updates aren't possible anyways. I changed it to use `IORef`s in c467874538e15340aca1e8a21c0ad7c6af26c93a.